### PR TITLE
[JCLOUDS-1357] This is the first commit of Softlayer Network API

### PR DESCRIPTION
**Pull request by swaqos**
Jira link: https://issues.apache.org/jira/browse/JCLOUDS-1357
JCLOUDS-1357
The update focuses on adding SoftLayer Network API to the features list of SoftLayer provider.

Domain type files:
- Added Network.java
- Added Subnet.java

Features:
- Added NetworkApi.java
- Added NetworkApiLiveTest.java
- Added NetworkApiMockTest.java
- Added resource files for NetworkApiMockTest.java

### CODE REVIEW

**Comment by andreaturli** [comment]
hi @swaqos thanks for your contribution!

Looks like the failures are related to this PR: some checkstyle violations -- see https://jclouds.ci.cloudbees.com/job/jclouds-pull-requests/2778/org.apache.jclouds.provider$softlayer/console for more details

Please have a look at `Coding standard` notes at https://cwiki.apache.org/confluence/display/JCLOUDS/Coding+Standards on how to fix those kind problems.

Thanks!

**Comment by swaqos** [comment]
Hi @andreaturli , thanks for the information. I updated the code and the build passed.

**Comment by andreaturli** [comment]
Thanks @swaqos I'll have a proper review shortly! thanks!

File: providers/softlayer/src/main/java/org/jclouds/softlayer/SoftLayerApi.java [hunk 1]
```diff
@@ -56,4 +57,11 @@
     */
    @Delegate
    AccountApi getAccountApi();
+
```
Comment (andreaturli) [inline providers/softlayer/src/main/java/org/jclouds/softlayer/SoftLayerApi.java]: [minor] remove extra un-needed line

File: providers/softlayer/src/main/java/org/jclouds/softlayer/domain/Network.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+package org.jclouds.softlayer.domain;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableList;
+import org.jclouds.javax.annotation.Nullable;
+import org.jclouds.json.SerializedNames;
+
+import java.util.List;
+
+// packets for auto value
```
Comment (andreaturli) [inline providers/softlayer/src/main/java/org/jclouds/softlayer/domain/Network.java]: [minor] remove this comment, as it doesn't seem relevant ;)

File: providers/softlayer/src/main/java/org/jclouds/softlayer/domain/Network.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+   public abstract int cidr();
+   public abstract String networkIdentifier();
+   @Nullable
+   public abstract String name();
+   @Nullable
+   public abstract String notes();
+   @Nullable
+   public abstract List<Subnet> subnets();
+
+
```
Comment (andreaturli) [inline providers/softlayer/src/main/java/org/jclouds/softlayer/domain/Network.java]: [minor] remove extra empty line

File: providers/softlayer/src/main/java/org/jclouds/softlayer/domain/Subnet.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+ */
+package org.jclouds.softlayer.domain;
+
+import com.google.auto.value.AutoValue;
+import org.jclouds.javax.annotation.Nullable;
+import org.jclouds.json.SerializedNames;
+
+import java.util.Date;
+
+// packets for auto value
```
Comment (andreaturli) [inline providers/softlayer/src/main/java/org/jclouds/softlayer/domain/Subnet.java]: [minor] remove this comment

File: providers/softlayer/src/test/java/org/jclouds/softlayer/features/NetworkApiLiveTest.java [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+/**
+ * Tests behavior of {@code NetworkApi}
+ */
+@Test(groups = "live")
+public class NetworkApiLiveTest extends BaseSoftLayerApiLiveTest {
+
+   private NetworkApi networkApi;
+
+   private Network network = null;
+   private List<Network> networks = null;
```
Comment (andreaturli) [inline providers/softlayer/src/test/java/org/jclouds/softlayer/features/NetworkApiLiveTest.java]: doesn't seem to be used, remove?

File: providers/softlayer/src/main/java/org/jclouds/softlayer/features/NetworkApi.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+import java.io.Closeable;
+import java.util.List;
+
+/**
+ * Provides access to Network via their REST API.
+ * <p/>
+ * @see <a href="http://sldn.softlayer.com/article/REST" />
+ */
+@RequestFilters(BasicAuthentication.class)
+@Path("/v3")
```
Comment (andreaturli) [inline providers/softlayer/src/main/java/org/jclouds/softlayer/features/NetworkApi.java]: I think we should use `@Path("/v{jclouds.api-version}")`

we may need to bump the `api-version` in `org.jclouds.softlayer.SoftLayerApiMetadata` -- in particular from `3` to `3.1` in https://github.com/jclouds/jclouds/blob/master/providers/softlayer/src/main/java/org/jclouds/softlayer/SoftLayerApiMetadata.java#L65, I think

File: providers/softlayer/src/test/java/org/jclouds/softlayer/features/NetworkApiLiveTest.java [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+   @BeforeClass(groups = {"integration", "live"})
+   @Override
+   public void setup() {
+      super.setup();
+
+      networkApi = api.getNetworkApi();
+
+      for (Datacenter dc : api.getDatacenterApi().listDatacenters()) {
+         if (dc.getName().equals("dal13")) {
```
Comment (andreaturli) [inline providers/softlayer/src/test/java/org/jclouds/softlayer/features/NetworkApiLiveTest.java]: why `dal013`? is the only datacenter that supports this network feature so far?
maybe consider making it a constant

File: providers/softlayer/src/test/java/org/jclouds/softlayer/features/NetworkApiLiveTest.java [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+
+   @Test(dependsOnMethods = "testCreateNetwork")
+   public void testGetNetwork() throws Exception {
+      Network found = networkApi.getNetwork(network.id());
+      assertEquals(found, network);
+      checkNetwork(found);
+      assertEquals(found.networkIdentifier(), "192.168.253.0");
+      assertEquals(found.name(), "NetworkApiLiveTestNetwork");
+      assertEquals(found.cidr(), 24);
+      assertEquals(found.notes(), "this is a test network from JClouds NetworkApiLiveTest.");
```
Comment (andreaturli) [inline providers/softlayer/src/test/java/org/jclouds/softlayer/features/NetworkApiLiveTest.java]: [minor] it's `jclouds`

File: providers/softlayer/src/test/java/org/jclouds/softlayer/features/NetworkApiLiveTest.java [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.assertTrue;
+import static org.testng.Assert.assertNull;
+
+/**
+ * Tests behavior of {@code NetworkApi}
+ */
+@Test(groups = "live")
+public class NetworkApiLiveTest extends BaseSoftLayerApiLiveTest {
```
Comment (andreaturli) [inline providers/softlayer/src/test/java/org/jclouds/softlayer/features/NetworkApiLiveTest.java]: Unfortunately my account is not allowed to create networks so I couldn't run the `NetworkApiLiveTest` -- `org.jclouds.http.HttpResponseException: command: POST https://api.softlayer.com/rest/v3.1/SoftLayer_Network/createObject HTTP/1.1 failed with response: HTTP/1.1 500 Internal Server Error; content: [{"error":"Cloudsoft Corporation (#xxxxx) is not allowed to create networks.","code":"SoftLayer_Exception_Account_Limit"}]` so could you maybe paste here the results from your account, maybe? thanks!

File: providers/softlayer/src/test/java/org/jclouds/softlayer/features/NetworkApiMockTest.java [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+import java.util.List;
+
+import static com.google.common.collect.Iterables.size;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+/**
+ * Mock tests for the {@link NetworkApi} class.
+ */
+@Test(groups = "unit", testName = "SecurityGroupApiMockTest")
+public class NetworkApiMockTest extends BaseSoftLayerMockTest {
```
Comment (andreaturli) [inline providers/softlayer/src/test/java/org/jclouds/softlayer/features/NetworkApiMockTest.java]: great start @swaqos!

Similarly to other provider, jclouds also tests fallbacks strategies when the operations are not successful.
For example, `listNetworks` is annotated with `@Fallback(Fallbacks.EmptyListOnNotFoundOr404.class)` so a test similar to https://github.com/jclouds/jclouds/blob/master/providers/packet/src/test/java/org/jclouds/packet/features/FacilityApiMockTest.java#L45 is required.

Please add the missing "bad case" tests for the operations in this MockTest class to improve the coverage

**Comment by andreaturli** [comment]
@swaqos if you name the PR with the jira issue, JCLOUDS-1357 it will be automatically linked to the jira issue.

File: providers/softlayer/src/test/java/org/jclouds/softlayer/features/NetworkApiLiveTest.java (hunk 9 above)
Comment (swaqos) [reply providers/softlayer/src/test/java/org/jclouds/softlayer/features/NetworkApiLiveTest.java]: Hi Andrea, I believe that only certain type of accounts have access to the new SL+ features.
The following is a mvn test report results:

```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=256m; support was removed in 8.0
Running org.jclouds.softlayer.features.NetworkApiLiveTest
Configuring TestNG with: TestNG652Configurator
Starting test testCreateNetwork(org.jclouds.softlayer.features.NetworkApiLiveTest)
[TestNG] Test testCreateNetwork(org.jclouds.softlayer.features.NetworkApiLiveTest) succeeded: 1710ms
Test suite progress: tests succeeded: 1, failed: 0, skipped: 0.
Starting test testCreateSubnet(org.jclouds.softlayer.features.NetworkApiLiveTest)
[TestNG] Test testCreateSubnet(org.jclouds.softlayer.features.NetworkApiLiveTest) succeeded: 3401ms
Test suite progress: tests succeeded: 2, failed: 0, skipped: 0.
Starting test testGetName(org.jclouds.softlayer.features.NetworkApiLiveTest)
[TestNG] Test testGetName(org.jclouds.softlayer.features.NetworkApiLiveTest) succeeded: 1206ms
Test suite progress: tests succeeded: 3, failed: 0, skipped: 0.
Starting test testGetNetwork(org.jclouds.softlayer.features.NetworkApiLiveTest)
[TestNG] Test testGetNetwork(org.jclouds.softlayer.features.NetworkApiLiveTest) succeeded: 1325ms
Test suite progress: tests succeeded: 4, failed: 0, skipped: 0.
Starting test testGetNotes(org.jclouds.softlayer.features.NetworkApiLiveTest)
[TestNG] Test testGetNotes(org.jclouds.softlayer.features.NetworkApiLiveTest) succeeded: 1147ms
Test suite progress: tests succeeded: 5, failed: 0, skipped: 0.
Starting test testListNetworks(org.jclouds.softlayer.features.NetworkApiLiveTest)
[TestNG] Test testListNetworks(org.jclouds.softlayer.features.NetworkApiLiveTest) succeeded: 7443ms
Test suite progress: tests succeeded: 6, failed: 0, skipped: 0.
Starting test testEditNetwork(org.jclouds.softlayer.features.NetworkApiLiveTest)
[TestNG] Test testEditNetwork(org.jclouds.softlayer.features.NetworkApiLiveTest) succeeded: 2570ms
Test suite progress: tests succeeded: 7, failed: 0, skipped: 0.
Starting test testGetSubnets(org.jclouds.softlayer.features.NetworkApiLiveTest)
[TestNG] Test testGetSubnets(org.jclouds.softlayer.features.NetworkApiLiveTest) succeeded: 1108ms
Test suite progress: tests succeeded: 8, failed: 0, skipped: 0.
Starting test testDeleteSubnet(org.jclouds.softlayer.features.NetworkApiLiveTest)
[TestNG] Test testDeleteSubnet(org.jclouds.softlayer.features.NetworkApiLiveTest) succeeded: 2658ms
Test suite progress: tests succeeded: 9, failed: 0, skipped: 0.
Starting test testDeleteNetwork(org.jclouds.softlayer.features.NetworkApiLiveTest)
[TestNG] Test testDeleteNetwork(org.jclouds.softlayer.features.NetworkApiLiveTest) succeeded: 11568ms
Test suite progress: tests succeeded: 10, failed: 0, skipped: 0.
Tests run: 10, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 53.068 sec - in org.jclouds.softlayer.features.NetworkApiLiveTest

Results :

Tests run: 10, Failures: 0, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 02:05 min
[INFO] Finished at: 2017-11-21T18:07:17+08:00
[INFO] Final Memory: 27M/330M
[INFO] ------------------------------------------------------------------------
```

**Comment by andreaturli** [comment]
Merged at [master](http://git-wip-us.apache.org/repos/asf/jclouds/commit/a540daf3)

Thanks again @swaqos -- looking forward to seeing more of your work!

**Pull request closed without merging**